mxnet/mxnet-engine/src/main/java/ai/djl/mxnet/engine/MxSymbolBlock.java

```diff
@@ -164,24 +165,24 @@ public NDList forward(
             NDList inputs,
             boolean training,
             PairList<String, Object> params) {
-        if (first) {
-            synchronized (MxSymbolBlock.class) {
-                if (first) {
-                    // create CachedOp is not thread-safe
-                    // add synchronized block to avoid creating multiple CachedOps
-                    op = JnaUtils.createCachedOp(this, (MxNDManager) manager, training);
-                    inputDescriptions = new PairList<>();
-                    outputDescriptions = new PairList<>();
-                    for (NDArray array : inputs) {
-                        inputDescriptions.add(array.getName(), array.getShape());
-                    }
-                    NDList outputs = op.forward(parameterStore, inputs, training);
-                    for (NDArray array : outputs) {
-                        outputDescriptions.add(array.getName(), array.getShape());
-                    }
-                    first = false;
-                    return outputs;
+        CachedOp op = opHolder.get();
+        if (op == null) {
+            op = JnaUtils.createCachedOp(this, (MxNDManager) manager, training);
```

Comment:
We need to move createCachedOp into synchronized (MxSymbolBlock.class) block as createCachedOp is not thread-safe

Reply:
opHolder is a threadLocal object, it will ensure the op got only applies to the specific thread